src/include/duckdb/common/insertion_order_preserving_map.hpp

```diff
@@ -0,0 +1,39 @@
+//===----------------------------------------------------------------------===//
+//                         DuckDB
+//
+// duckdb/common/insertion_order_preserving_map.hpp
+//
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "duckdb/common/unordered_map.hpp"
+#include "duckdb/common/unordered_set.hpp"
+#include "duckdb/common/string.hpp"
+#include "duckdb/common/string_util.hpp"
+#include "duckdb/common/helper.hpp"
+#include "duckdb/common/case_insensitive_map.hpp"
+
+namespace duckdb {
+
+template <typename K, typename V>
+class InsertionOrderPreservingMap : public vector<V> {
```

Comment:
Inheriting from `vector` seems like it is a bit error-prone here, since the `map_idx` and the vector could go out-of-sync quite easily (there is an invariant that `map_idx.size() == vector.size()`, but that is not enforced by the API here).

Perhaps we can instead have two private member variables that represent the map/vector, and implement the relevant methods that are necessary for the CTE map so that the invariant is maintained?

I think a nice interface here would be if the `InsertionOrderPreservingMap` behaves like a `std::map` in that we can access members using `operator[]`, or `.find()`, but when we iterate over the map (i.e. using `for(auto &entry : map)`) the entries are iterated over in insertion-order.

`K` should also always be an `idx_t`, so we should not need to parameterize that, right? Currently the actual key is always a string (as part of `case_insensitive_map_t`).

Reply:
Great idea! You are right, the map and vector should never get out of sync. An explicit interface should be the way to go here.